Psalm integration

weierophinney opened the pull request:
Adds Psalm integration to the CI workflow:

- Adds a `psalm.xml.dist` configuration file and `psalm-baseline.xml`
- Requires vimeo/psalm and psalm/plugin-phpunit as dev dependencies
- Removes phpspec/prophecy as a dev dependency
- Bumps PHPUnit version to 9.3 as minimum
- Adds static-analysis script
- Updates Travis config to run static analysis on test coverage targets
- Performs a number of changes suggested by Psalm where they were low-hanging fruit

Fixes #7

weierophinney committed:
feat: add Psalm integration

- Adds dev requirements on vimeo/psalm and psalm/plugin-phpunit
- Adds static-analysis script
- Updates Travis config to run psalm when test-coverage runs

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: incorporate low-hanging fruit fixes from Psalm baseline report

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: remove prophecy, and update phpunit minimum to 9.3

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

lcobucci commented on test/ConfigResourceFactoryTest.php:
```diff
@@ -10,19 +10,19 @@
 {
-    use ProphecyTrait;
+    /** @var string */
+    private const WRITER_SERVICE = 'Laminas\ApiTools\Configuration\ConfigWriter';
```
Could this use the `::class` syntax?

weierophinney replied:
Psalm was complaining about it, as the class doesn't actually exist; it's a pseudo-class for specifying a config writer for the container.

lcobucci commented on test/ConfigResourceFactoryTest.php:
```diff
@@ -32,44 +32,54 @@ class ConfigResourceFactoryTest extends TestCase
+    public function testDefaultAttributesValues(): void
     {
-        $this->container->has('config')->willReturn(false);
+        $this->container->expects($this->atLeastOnce())->method('has')->with('config')->willReturn(false);
```
This could be used to keep the stub behaviour (not caring about number of calls nor arguments):


```suggestion
        $this->container->method('has')->willReturn(false);
```

lcobucci commented on test/ConfigResourceFactoryTest.php:
```diff
@@ -32,44 +32,54 @@ class ConfigResourceFactoryTest extends TestCase
-        $this->container->has('config')->willReturn(false);
+        $this->container->expects($this->atLeastOnce())->method('has')->with('config')->willReturn(false);
+        $this->container
+            ->expects($this->once())
```
Once is a static method (at least in the recent phpunit versions) 

```suggestion
            ->expects(self::once())
```

lcobucci commented on test/ResourceFactoryTest.php:
```diff
@@ -15,20 +15,23 @@
     {
         $this->resourceFactory = new ResourceFactory(
-            $this->getMockBuilder(ModuleUtils::class, [], [], '', false)
+            $this->getMockBuilder(ModuleUtils::class)
```
AFAIK that's the default behaviour of `$this->createMock()`

weierophinney replied:
Recent versions (and the ones allowed with current commits on this branch) only have a single argument; Psalm had flagged that for me, which is why I made the change.

weierophinney replied:
Ah, missed that you were talking `createMock()`, vs `getMockBuilder()`; updated!

lcobucci reviewed (approved):
Just some teeny tiny optional things. The stub vs mock argument is applicable everywhere, btw.

weierophinney committed:
qa: call methods statically when defined statically

`once()`, `atLeastOnce()`, `returnValueMap()` are all defined as static methods.

Also, no reason to use `atLeastOnce()` when we just want to ensure it returns what is expected.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: use `createMock()` instead of `getMockBuilder()` + `disableOriginalConstructor()`

It's the same behavior.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
